File: Nudge-Neurodivergent-task-tracker/RoutineManager.java
import java.time.LocalTime;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

/**
 * Manages morning, evening, and custom routines with flexible scheduling
 */
public class RoutineManager {
    public enum RoutineType {
        MORNING("🌅 Morning Routine"),
        EVENING("🌙 Evening Routine"),
        WORK_START("💼 Work Start Routine"),
        WORK_END("🏠 Work End Routine"),
        CUSTOM("⭐ Custom Routine");
        
        private final String displayName;
        
        RoutineType(String displayName) {
            this.displayName = displayName;
        }
        
        public String getDisplayName() { return displayName; }
    }
    
    public static class RoutineStep {
        private String description;
        private int estimatedMinutes;
        private boolean isCompleted;
        private boolean isOptional;
        
        public RoutineStep(String description, int estimatedMinutes, boolean isOptional) {
            this.description = description;
            this.estimatedMinutes = estimatedMinutes;
            this.isOptional = isOptional;
            this.isCompleted = false;
        }
        
        public void markCompleted() { this.isCompleted = true; }
        public void reset() { this.isCompleted = false; }
        
        // Getters
        public String getDescription() { return description; }
        public int getEstimatedMinutes() { return estimatedMinutes; }
        public boolean isCompleted() { return isCompleted; }
        public boolean isOptional() { return isOptional; }
        
        @Override
        public String toString() {
            String status = isCompleted ? "✅" : "⬜";
            String optionalMark = isOptional ? " (optional)" : "";
            return String.format("%s %s - %d min%s", status, description, estimatedMinutes, optionalMark);
        }
    }
    
    public static class Routine {
        private String name;
        private RoutineType type;
        private LocalTime preferredTime;
        private List<RoutineStep> steps;
        private boolean isActive;
        private int completionCount;
        
        public Routine(String name, RoutineType type, LocalTime preferredTime) {
            this.name = name;
            this.type = type;
            this.preferredTime = preferredTime;
            this.steps = new ArrayList<>();
            this.isActive = true;
            this.completionCount = 0;
        }
        
        public void addStep(String description, int minutes, boolean isOptional) {
            steps.add(new RoutineStep(description, minutes, isOptional));
        }
        
        public void removeStep(int index) {
            if (index >= 0 && index < steps.size()) {
                steps.remove(index);
            }
        }
        
        public void completeStep(int index) {
            if (index >= 0 && index < steps.size()) {
                steps.get(index).markCompleted();
            }
        }
        
        public void resetRoutine() {
            steps.forEach(RoutineStep::reset);
        }
        
        public boolean isFullyCompleted() {
            return steps.stream()
                .filter(step -> !step.isOptional())
                .allMatch(RoutineStep::isCompleted);
        }
        
        public int getCompletionPercentage() {
            if (steps.isEmpty()) return 100;
            
            long completed = steps.stream()
                .filter(step -> !step.isOptional())
                .mapToInt(step -> step.isCompleted() ? 1 : 0)
                .sum();
            
            long total = steps.stream()
                .mapToInt(step -> step.isOptional() ? 0 : 1)
                .sum();
            
            return total == 0 ? 100 : (int) ((completed * 100) / total);
        }
        
        public int getTotalEstimatedTime() {
            return steps.stream().mapToInt(RoutineStep::getEstimatedMinutes).sum();
        }
        
        public void markCompleted() {
            completionCount++;
            resetRoutine();
        }
        
        public void displayRoutine() {
            System.out.println("\n" + type.getDisplayName() + " - " + name);
            System.out.println("⏰ Preferred time: " + preferredTime);
            System.out.println("📊 Completion: " + getCompletionPercentage() + "%");
            System.out.println("⏱️ Total time: ~" + getTotalEstimatedTime() + " minutes");
            System.out.println("-".repeat(40));
            
            for (int i = 0; i < steps.size(); i++) {
                System.out.printf("%d. %s\n", i + 1, steps.get(i));
            }
        }
        
        // Getters and setters
        public String getName() { return name; }
        public RoutineType getType() { return type; }
        public LocalTime getPreferredTime() { return preferredTime; }
        public List<RoutineStep> getSteps() { return new ArrayList<>(steps); }
        public boolean isActive() { return isActive; }
        public int getCompletionCount() { return completionCount; }
        
        public void setActive(boolean active) { this.isActive = active; }
        public void setPreferredTime(LocalTime time) { this.preferredTime = time; }
    }
    
    private Map<RoutineType, List<Routine>> routines;
    
    public RoutineManager() {
        this.routines = new HashMap<>();
        for (RoutineType type : RoutineType.values()) {
            routines.put(type, new ArrayList<>());
        }
        
        createDefaultRoutines();
    }
    
    private void createDefaultRoutines() {
        // Default morning routine
        Routine morningRoutine = new Routine("Basic Morning", RoutineType.MORNING, LocalTime.of(8, 0));
        morningRoutine.addStep("💊 Take medication", 2, false);
        morningRoutine.addStep("🚿 Shower or wash face", 10, false);
        morningRoutine.addStep("🥣 Eat breakfast", 15, false);
        morningRoutine.addStep("📋 Review today's priorities", 5, true);
        morningRoutine.addStep("🧘 5-minute mindfulness", 5, true);
        addRoutine(morningRoutine);
        
        // Default evening routine
        Routine eveningRoutine = new Routine("Wind Down", RoutineType.EVENING, LocalTime.of(21, 0));
        eveningRoutine.addStep("📱 Put devices away", 2, false);
        eveningRoutine.addStep("🦷 Brush teeth", 3, false);
        eveningRoutine.addStep("📝 Tomorrow's top 3 tasks", 5, true);
        eveningRoutine.addStep("📖 Read or listen to calming audio", 20, true);
        eveningRoutine.addStep("😴 Prepare sleep environment", 5, false);
        addRoutine(eveningRoutine);
    }
    
    public void addRoutine(Routine routine) {
        routines.get(routine.getType()).add(routine);
    }
    
    public void removeRoutine(RoutineType type, String name) {
        routines.get(type).removeIf(routine -> routine.getName().equals(name));
    }
    
    public List<Routine> getRoutinesByType(RoutineType type) {
        return new ArrayList<>(routines.get(type));
    }
    
    public List<Routine> getAllActiveRoutines() {
        List<Routine> allActive = new ArrayList<>();
        for (List<Routine> typeRoutines : routines.values()) {
            typeRoutines.stream()
                .filter(Routine::isActive)
                .forEach(allActive::add);
        }
        return allActive;
    }
    
    public Routine findRoutine(RoutineType type, String name) {
        return routines.get(type).stream()
            .filter(routine -> routine.getName().equals(name))
            .findFirst()
            .orElse(null);
    }
    
    public List<String> getRoutineReminders() {
        List<String> reminders = new ArrayList<>();
        LocalTime now = LocalTime.now();
        
        for (Routine routine : getAllActiveRoutines()) {
            LocalTime routineTime = routine.getPreferredTime();
            
            // Check if routine time is within 30 minutes
            if (Math.abs(now.toSecondOfDay() - routineTime.toSecondOfDay()) <= 1800) {
                if (routine.getCompletionPercentage() < 100) {
                    reminders.add("⏰ " + routine.getType().getDisplayName() + 
                                " time! (" + routine.getName() + ")");
                }
            }
        }
        
        return reminders;
    }
    
    public void displayAllRoutines() {
        System.out.println("\n📋 YOUR ROUTINES");
        System.out.println("=".repeat(50));
        
        for (RoutineType type : RoutineType.values()) {
            List<Routine> typeRoutines = routines.get(type);
            if (!typeRoutines.isEmpty()) {
                System.out.println("\n" + type.getDisplayName() + ":");
                for (Routine routine : typeRoutines) {
                    if (routine.isActive()) {
                        System.out.printf("  📌 %s - %d%% complete (%d times done)\n", 
                                        routine.getName(), 
                                        routine.getCompletionPercentage(),
                                        routine.getCompletionCount());
                    }
                }
            }
        }
    }
    
    public String getRoutineStats() {
        int totalRoutines = (int) getAllActiveRoutines().size();
        int completedToday = (int) getAllActiveRoutines().stream()
            .mapToInt(Routine::getCompletionPercentage)
            .filter(percentage -> percentage == 100)
            .count();
        
        return String.format("📊 Routines: %d/%d completed today", completedToday, totalRoutines);
    }
    
    public List<String> getRoutineSuggestions(EnergyTracker.EnergyLevel energy) {
        List<String> suggestions = new ArrayList<>();
        
        switch (energy) {
            case VERY_LOW:
                suggestions.add("🌸 Keep routines super minimal today");
                suggestions.add("🛁 Focus on just the self-care essentials");
                suggestions.add("💊 Don't skip medications though!");
                break;
            case LOW:
                suggestions.add("📝 Pick just 2-3 routine items that matter most");
                suggestions.add("⏰ Maybe do them slower than usual");
                suggestions.add("🤗 Be extra gentle with yourself");
                break;
            case GOOD:
            case HIGH:
                suggestions.add("✨ Great day to tackle full routines!");
                suggestions.add("➕ Maybe add an optional step you've been skipping");
                suggestions.add("🎯 Consider optimizing your routine flow");
                break;
        }
        
        return suggestions;
    }
}

File: Nudge-Neurodivergent-task-tracker/routine_manager.py
"""
Routine management for morning and evening routines
"""

from datetime import datetime, time
from enum import Enum
from typing import List, Dict
from dataclasses import dataclass
from energy_tracker import EnergyLevel


class RoutineType(Enum):
    """Types of routines"""
    MORNING = "Morning"
    EVENING = "Evening"
    CUSTOM = "Custom"


@dataclass
class RoutineStep:
    """A single step in a routine"""
    name: str
    description: str
    estimated_minutes: int
    completed: bool = False


@dataclass
class Routine:
    """A routine with multiple steps"""
    name: str
    routine_type: RoutineType
    steps: List[RoutineStep]
    completion_count: int = 0
    
    def display_routine(self):
        """Display the routine with progress"""
        print(f"\n📅 {self.name.upper()}")
        print("-" * 40)
        completed_steps = sum(1 for step in self.steps if step.completed)
        total_time = sum(step.estimated_minutes for step in self.steps)
        
        print(f"Progress: {completed_steps}/{len(self.steps)} steps")
        print(f"Estimated time: {total_time} minutes")
        print("\nSteps:")
        
        for i, step in enumerate(self.steps, 1):
            status = "✅" if step.completed else "⚪"
            print(f"  {status} {i}. {step.name} ({step.estimated_minutes}min)")
            if step.description:
                print(f"      💡 {step.description}")
    
    def complete_step(self, step_index: int) -> bool:
        """Complete a step in the routine"""
        if 0 <= step_index < len(self.steps):
            self.steps[step_index].completed = True
            return True
        return False
    
    def reset_routine(self):
        """Reset all steps to incomplete"""
        for step in self.steps:
            step.completed = False
    
    def is_completed(self) -> bool:
        """Check if all steps are completed"""
        return all(step.completed for step in self.steps)


class RoutineManager:
    """Manages daily routines"""
    
    def __init__(self):
        self.routines: List[Routine] = []
        self._create_default_routines()
    
    def _create_default_routines(self):
        """Create default morning and evening routines"""
        # Morning routine
        morning_steps = [
            RoutineStep("Take medication", "If applicable", 2),
            RoutineStep("Drink water", "Hydrate after sleep", 2),
            RoutineStep("Check energy level", "Quick self-assessment", 1),
            RoutineStep("Review today's priorities", "3-5 main things", 5),
            RoutineStep("Breakfast/nutrition", "Fuel your brain", 15),
            RoutineStep("Quick self-care", "Hygiene, comfortable clothes", 15)
        ]
        
        morning_routine = Routine(
            name="Basic Morning",
            routine_type=RoutineType.MORNING,
            steps=morning_steps
        )
        
        # Evening routine
        evening_steps = [
            RoutineStep("Reflect on day", "What went well?", 5),
            RoutineStep("Tomorrow preparation", "Set out clothes, check calendar", 10),
            RoutineStep("Digital sunset", "Reduce screen time", 5),
            RoutineStep("Calming activity", "Reading, music, bath", 20),
            RoutineStep("Gratitude practice", "3 things you're grateful for", 3),
            RoutineStep("Sleep preparation", "Comfortable environment", 10)
        ]
        
        evening_routine = Routine(
            name="Wind Down",
            routine_type=RoutineType.EVENING,
            steps=evening_steps
        )
        
        self.routines = [morning_routine, evening_routine]
    
    def get_routine_reminders(self) -> List[str]:
        """Get routine reminders based on time of day"""
        current_hour = datetime.now().hour
        reminders = []
        
        if 6 <= current_hour <= 10:  # Morning
            morning_routines = [r for r in self.routines if r.routine_type == RoutineType.MORNING]
            for routine in morning_routines:
                if not routine.is_completed():
                    incomplete_steps = [s for s in routine.steps if not s.completed]
                    if incomplete_steps:
                        next_step = incomplete_steps[0]
                        reminders.append(f"🌅 Morning: {next_step.name}")
        
        elif 18 <= current_hour <= 23:  # Evening
            evening_routines = [r for r in self.routines if r.routine_type == RoutineType.EVENING]
            for routine in evening_routines:
                if not routine.is_completed():
                    incomplete_steps = [s for s in routine.steps if not s.completed]
                    if incomplete_steps:
                        next_step = incomplete_steps[0]
                        reminders.append(f"🌙 Evening: {next_step.name}")
        
        return reminders
    
    def get_routine_suggestions(self, energy_level: EnergyLevel) -> List[str]:
        """Get routine suggestions based on energy level"""
        suggestions = []
        
        if energy_level.level_value <= 2:  # Low energy
            suggestions.extend([
                "🌸 Keep morning routine minimal - just essentials",
                "☕ Focus on hydration and basic nutrition",
                "🛁 Prioritize comfort and gentle self-care",
                "📱 Consider voice memos instead of writing"
            ])
        elif energy_level.level_value >= 4:  # High energy
            suggestions.extend([
                "🚀 Great time to establish new routine habits",
                "📋 Consider adding planning/prep steps",
                "💪 Use this energy for routine optimization",
                "⏰ Maybe tackle some weekly prep tasks"
            ])
        else:  # Moderate energy
            suggestions.extend([
                "⚖️ Standard routines should feel manageable",
                "🎯 Focus on consistency over perfection",
                "📝 Good time for routine evaluation",
                "🔄 Consider small routine improvements"
            ])
        
        return suggestions
    
    def display_all_routines(self):
        """Display all available routines"""
        print("\n📋 YOUR ROUTINES")
        print("=" * 50)
        
        for routine in self.routines:
            completed_steps = sum(1 for step in routine.steps if step.completed)
            completion_percent = (completed_steps / len(routine.steps)) * 100
            
            routine_icon = "🌅" if routine.routine_type == RoutineType.MORNING else "🌙"
            print(f"{routine_icon} {routine.routine_type.value} Routine:")
            print(f"  📌 {routine.name} - {completion_percent:.0f}% complete ({routine.completion_count} times done)")
    
    def get_all_active_routines(self) -> List[Routine]:
        """Get all active routines"""
        return self.routines.copy()
    
    def get_routine_stats(self) -> str:
        """Get routine completion statistics"""
        total_routines = len(self.routines)
        total_completions = sum(r.completion_count for r in self.routines)
        
        return f"📅 Routines: {total_routines} active, {total_completions} total completions"
    
    def add_custom_routine(self, name: str, steps: List[RoutineStep]):
        """Add a custom routine"""
        custom_routine = Routine(
            name=name,
            routine_type=RoutineType.CUSTOM,
            steps=steps
        )
        self.routines.append(custom_routine)
        return custom_routine